asmcomp/cmmgen.ml

```diff
@@ -2434,6 +2434,55 @@ and transl_letrec env bindings cont =
         fill_blocks rem
   in init_blocks bsz
 
+(* Insert instrumentation for afl-fuzz *)
+
+let afl_area_ptr = Cconst_symbol "caml_afl_area_ptr"
+let afl_prev_loc = Cconst_symbol "caml_afl_prev_loc"
+let afl_map_size = 1 lsl 16
+
+let rec with_afl_logging b =
+  if !Clflags.afl_inst_ratio < 100 && Random.int 100 >= !Clflags.afl_inst_ratio then instrument b else
+  let instrumentation =
+    let cur_location = Random.int afl_map_size in
+    let cur_pos = Ident.create "pos" in
+    let afl_area = Ident.create "shared_mem" in
+    Clet(afl_area, Cop(Cload Word_int, [afl_area_ptr]),
+    Clet(cur_pos,  Cop(Cxor, [Cop(Cload Word_int, [afl_prev_loc]); Cconst_int cur_location]),
+    Csequence(
+      Cop(Cstore(Byte_unsigned, Assignment),
+          [Cop(Cadda, [Cvar afl_area; Cvar cur_pos]);
+           Cop(Cadda, [Cop (Cload Byte_unsigned,
+                            [Cop(Cadda, [Cvar afl_area; Cvar cur_pos])]);
+                       Cconst_int 1])]),
+      Cop(Cstore(Word_int, Assignment),
+          [afl_prev_loc; Cconst_int (cur_location lsr 1)])))) in
+  Csequence(instrumentation, instrument b)
+
+and instrument = function
+  (* these cases add afl logging, since they may be targets of conditional branches *)
+  | Cifthenelse (cond, t, f) -> Cifthenelse (instrument cond, with_afl_logging t, with_afl_logging f)
+  | Ccatch (nfail, ids, e1, e2) ->
+     Ccatch (nfail, ids, instrument e1, with_afl_logging e2)
```

Comment:
This probably don't need logging as the relevant branching was normally already reported before going to the Cexit